### ISSUE TITLE
ansible should warn the user if absense of the sftp subsystem prevents normal operation

### ISSUE THREAD

**Issue by wu-lee** [ansible/ansible#9332]
##### Issue Type:

Feature Idea
##### Ansible Version:

ansible --version
ansible 1.8 (devel 2eda9a3a47) last updated 2014/10/10 18:43:36 (GMT +100)
  lib/ansible/modules/core: (detached HEAD 5af8d55b03) last updated 2014/10/10 18:43:44 (GMT +100)
  lib/ansible/modules/extras: (detached HEAD 681db4ce2c) last updated 2014/10/10 18:43:45 (GMT +100)
  v2/ansible/modules/core: (detached HEAD cb69744bce) last updated 2014/10/08 10:13:54 (GMT +100)
  v2/ansible/modules/extras: (detached HEAD 8a4f07eecd) last updated 2014/10/08 10:14:38 (GMT +100)
  configured module search path = /home/nick/i/gitworking/controllis/ansible/ansible/library:library/modules
##### Environment:

Ubuntu 12.04.5, connecting to CentOS 6.5
##### Summary:

Ansible connections suddently started failing to some, but nor all servers.  Further, somewhat laborious investigation discovered the cause: the sshd config had been updated, disabling the SFTP subsystem.

The problem: ansible was not at all helpful.  Please can ansible tell the user if a required subsystem is not available, otherwise this is a very hard problem to diagnose. See examples below.
##### Steps To Reproduce:

First, take a working server, and obviously check ansible can connect.  Also test sftp works:

```
$ sftp root@192.168.69.164
Connected to 192.168.69.164
```

Then disable the sftp subsystem on that server, in my case this happend by accident as a Debian style sshd_config had been installed on a Red Hat server, so this:

```
# override default of no subsystems
Subsystem   sftp    /usr/libexec/openssh/sftp-server
```

became:

```
# override default of no subsystems
Subsystem   sftp    /usr/lib/openssh/sftp-server
```

Now try sftp again:

```
$ sftp root@192.168.69.164
subsystem request failed on channel 0
Couldn't read packet: Connection reset by peer
```

However, ansible behaves like this:

```
$ ansible -u root 192.168.69.164 -m command -a 'ls /'  
zorro.controllis.com | FAILED => failed to transfer file to /root/.ansible/tmp/ansible-tmp-1413283186.93-222428902594155/command:

Couldn't read packet: Connection reset by peer
```

Note, no information why the connection was reset.  Even using -vvvv doesn't help (unless you happen to spot the clue from the word "PUT", which I didn't initially):

```
ansible -vvvv -u root 192.168.69.164 -m command -a 'ls /'  
<192.168.69.164> ESTABLISH CONNECTION FOR USER: root
<192.168.69.164> REMOTE_MODULE command ls /
<192.168.69.164> EXEC ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath=/home/nick/.ansible/cp/ansible-ssh-%h-%p-%r -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=10 192.168.69.164 /bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1413282510.89-11840126958552 && echo $HOME/.ansible/tmp/ansible-tmp-1413282510.89-11840126958552'
<192.168.69.164> PUT /tmp/tmpDIqUQC TO /root/.ansible/tmp/ansible-tmp-1413282510.89-11840126958552/command
zorro.controllis.com | FAILED => failed to transfer file to /root/.ansible/tmp/ansible-tmp-1413282510.89-11840126958552/command:

Couldn't read packet: Connection reset by peer
```

Change "command" to "raw" above, and it will work, because the initial 'setup' module is not invoked, which uses sftp by default.

I eventually ran strace on my sshd server and inspected the output, and this demonstrated that it was not in fact the initial ssh connection had failed.  Due to the persistent connection, I needed to delete my ~/.ansible directory to trigger a new connection, a further obstacle.

Setting scp_if_ssh = True in ansible.cfg resolves the problem simply, when you know what is is.
##### Expected Results:

It'd be much more helpful if ansible picked up the reason, as the sftp command does, and maybe suggested a fix:

```
$ ansible -u root 192.168.69.164 -m command -a 'ls /'  
zorro.controllis.com | FAILED => failed to transfer file to /root/.ansible/tmp/ansible-tmp-1413283186.93-222428902594155/command:

subsystem request failed on channel 0
Couldn't read packet: Connection reset by peer
Ansible requires the sftp subsystem by default; if this is unavailable you might try setting the 
scp_if_scp option to 'True' in your ansible configuration <link to docs>
```

**Comment by mpdehaan** [comment]
Thanks for the suggestion, you might also wish to consider in ansible.cfg switching to scp mode.

**Comment by wu-lee** [comment]
As mentioned, that works, the point I'm making is that you may not know it's the right thing to do.

**Comment by Gerst20051** [comment]
@wu-lee nice suggestion dude he should definitely incorporate this into ansible... maybe you can make a pull request for this :) :+1:

**Comment by benallard** [comment]
This would have saved me a few troubles. :+1: !

**Comment by maticmeznar** [comment]
I just had the same problem when I applied RedHat family sshd_config on Debian :+1:

**Comment by iliv** [comment]
In my case it was "Subsystem sftp sftp-internal" in /etc/ssh/sshd_config that broke ansible. 

It would error out with

> FAILED => failed to transfer file to /home/ansible/.ansible/tmp/ansible-tmp-1441383683.03- 253999378416460/command:
> 
> Connection closed

and took hours to identify the problem. I support this suggestion to make ansible print more informative errors when it encounters difficulties using SFTP. As a bonus, it should probably also suggest to use scp_if_ssh=True option to try to work around possible SFTP related issues.

**Comment by bcoca** [comment]
I just need to check if "subsystem request failed" is reliable way to detect this

**Comment by bcoca** [comment]
fixed via #18643, which adds a fallback method
